bluesky/callbacks/core.py

```diff
@@ -1,17 +1,45 @@
-"""
-Useful callbacks for the Run Engine
+""" Useful callbacks for the Run Engine
 """
 from itertools import count
 import warnings
-from collections import deque, namedtuple, OrderedDict
+from collections import defaultdict, deque, namedtuple, OrderedDict, ChainMap
 import time as ttime
-
 from datetime import datetime
 import logging
 from ..utils import ensure_uid
 
+from event_model import DocumentRouter
+
 
 class CallbackBase:
+    """
+    A convenient platform for writing stateful callbacks.
+
+    Examples
+    --------
+    A callback that prints the total number of events as they go by.
+
+    >>> class Thing(CallbackBase):
+    ...     def __init__(self, start_doc=None):
+    ...         super().__init__(start_doc)
+    ...         self.counter = 0
+    ...
+    ...     def event(self, doc):
+    ...         self.counter += 1       
+    ...         print(self.counter)
+    """
+
+    def __init__(self, start_doc=None):
+        # Callbacks meant to be used with a RunRouter may *require* a start_doc
+        # at __init__ time by making the start_doc argument required.
+        # Others may not.
+        #
+        # If a start_doc is given, it is passed to the `start` method
+        # immediately, here. If one is not given, it is up to the caller to
+        # call `start(...)` later once the RunStart document is available.
+        if start_doc is not None:
```

Comment:
Why is this needed? Shouldn't we this handled by the Callback factory explicitly (by calling `start` after the init).